README.md
# Static Site Generator App Extension for Quasar v2, the Vue.js Framework

> A [Quasar v2](https://quasar.dev/) App Extension to generate static site AKA [JAMstack](https://jamstack.org).

> :new: Supports Vite since [v4.4.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.3.0...v4.4.0).

![npm](https://img.shields.io/npm/v/quasar-app-extension-ssg) ![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/freddy38510/quasar-app-extension-ssg) ![GitHub repo size](https://img.shields.io/github/repo-size/freddy38510/quasar-app-extension-ssg) ![npm](https://img.shields.io/npm/dt/quasar-app-extension-ssg) [![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

This project was created to fill this [Feature Request](https://github.com/quasarframework/quasar/issues/2299) from Quasar.

:warning: Looking for [Quasar v1](https://v1.quasar.dev/) with Vue 2 ? See corresponding [quasar-app-extension-ssg v2](https://github.com/freddy38510/quasar-app-extension-ssg/tree/2.x) documentation instead of latest version.

[Live Demo](#live-demo) | [Installing](#installing) | [Uninstalling](#uninstalling) | [Upgrading](#upgrading) | [Developing](#developing) | [Usage](#usage) | [Configuration](#configuration) | [Infos](#infos)

## Live Demo

A live demo built from a fresh new [Quasar project (with Vite)](demos/quasar-ssg-vite) is available at [https://quasar-app-extension-ssg-vite.netlify.app](https://quasar-app-extension-ssg-vite.netlify.app/).

The demo achieves a [Google PageSpeed ​​Insights](https://pagespeed.web.dev) score of **100** for mobile and desktop platforms.

<details>
  <summary>View mobile report</summary>

[![mobile report](demos/quasar-ssg-vite/pagespeed-report-mobile.png)](https://pagespeed.web.dev/report?url=https%3A%2F%2Fquasar-app-extension-ssg-vite.netlify.app%2F&form_factor=mobile)

</details>

<details>
  <summary>View desktop report</summary>

[![desktop report](demos/quasar-ssg-vite/pagespeed-report-desktop.png)](https://pagespeed.web.dev/report?url=https%3A%2F%2Fquasar-app-extension-ssg-vite.netlify.app%2F&form_factor=desktop)

</details>

## Installing

Run this command into the Quasar project:

```bash
quasar ext add ssg
```

This will find and install the extension’s module. Once the installation is complete, interactive [prompts](#prompts) will wait for responses.

### Prompts

<details>
  <summary>Vite</summary>

- `add scripts into package.json?`: Extends package.json by adding scripts.

  ```javascript
  scripts: {
    'build:ssg': 'quasar ssg generate',
    'serve:ssg': 'quasar ssg serve dist/ssg'
  }
  ```

- `Add auto-completion of ssg property of quasar.config.js file for IDE ?`: Augments `configure()` helper from [`quasar/wrappers`](https://github.com/quasarframework/quasar/blob/quasar-v2.9.2/ui/wrappers/index.js#L17&L19) to get a better [IDE autocomplete](https://quasar.dev/quasar-cli-vite/quasar-config-js#ide-autocompletion) experience with the extension.

- `Inline critical css and async load the rest ?`: Uses [Beastcss](https://github.com/freddy38510/beastcss) to inline critical CSS and async load the rest for each generated page.
</details>

<details>
  <summary>Webpack</summary>

- `add scripts into package.json?`: Extends package.json by adding scripts.

  ```javascript
  scripts: {
    'build:ssg': 'quasar ssg generate',
    'serve:ssg': 'quasar ssg serve dist/ssg'
  }
  ```

- `Add auto-completion of ssg property of quasar.config.js file for IDE ?`: Augments `configure()` helper from [`quasar/wrappers`](https://github.com/quasarframework/quasar/blob/quasar-v2.9.2/ui/wrappers/index.js#L17&L19) to get a better [IDE autocomplete](https://quasar.dev/quasar-cli-vite/quasar-config-js#ide-autocompletion) experience with the extension.

- `Inline critical css and async load the rest ?`: Uses [Beastcss](https://github.com/freddy38510/beastcss) to inline critical CSS and async load the rest for each generated page.

- `Inline CSS from Vue SFC <style> blocks ?`: Inlines css from Vue Single-File Component (SFC) `<style>` blocks.
</details>

## Uninstalling

```bash
quasar ext remove ssg
```

## Upgrading

This is done with the same command used for installation:

```bash
quasar ext add ssg
```

## Developing

To develop the extension start by cloning this repository:

```bash
git clone https://github.com/freddy38510/quasar-app-extension-ssg.git && cd quasar-app-extension-ssg
```

Register the App Extension through yarn:

```bash
yarn link
```

Install dependencies:

```bash
yarn
```

Create a new Quasar project then install the App Extension:

```bash
yarn create quasar

cd <project-name>

quasar ext add ssg
```

Finally link the locally developed App Extension:

```bash
yarn link quasar-app-extension-ssg
```

Now the app extension can be developed without uninstalling/installing it every time something changes in it.

## Usage

### Generate

To generate a static site run this command from the quasar project folder:

```bash
quasar ssg generate
```

#### Generate Options

- `-h, --help`: Display usage instructions.
- `--force-build`: Force to build the application with webpack.
- `-d, --debug`: Build for debugging purposes.

### Dev

:new: _Added in [v4.2.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.1.1...v4.2.0)_

Starts the app in development mode (live reloading, error reporting, etc):

```bash
quasar ssg dev
```

The development server allows to develop the app by compiling and maintaining code in-memory. A web server will serve the app while offering live-reload out of the box. Running in-memory offers faster rebuilds when the code is changed.

The server can be configured by editing the `/quasar.config.js’ file:

```javascript
devServer: {
  host: '...',
  port: ...
}
```

#### Dev Options

- `-h, --help`: Display usage instructions.
- `--port, -p`: A port number on which to start the application.
- `--hostname, -H`: A hostname to use for serving the application.
- `--devtools, -d`: Open remote Vue Devtools.

### Serve

This extension provides a command to create a server to locally test the generated static site:

```bash
quasar ssg serve <dist-folder>
```

> Notes: This server is based on the [Quasar cli server](https://github.com/quasarframework/quasar/blob/dev/cli/bin/quasar-serve) adapted for static site. It serves the SPA fallback file (404.html) when a page has not been generated for a given route.

#### Serve Options

- `--port, -p`: Port to use (default: 4000).
- `--hostname, -H`: Address to use (default: 0.0.0.0).
- `--prefix-path`: Create a virtual path prefix (default: /).
- `--gzip, -g`: Compress content (default: true).
- `--silent, -s`: Suppress log message.
- `--colors`: Log messages with colors (default: true).
- `--open, -o`: Open browser window after starting.
- `--cache, -c <number>`: Cache time (max-age) in seconds. Does not apply to /service-worker.js (default: 86400 - 24 hours).
- `--micro, -m <seconds>`: Use micro-cache (default: 1 second).
- `--https`: Enable HTTPS.
- `--cert, -C [path]`: Path to SSL cert file (Optional).
- `--key, -K [path]`: Path to SSL key file (Optional).
- `--proxy <file.js>`: Proxy specific requests defined in file. File must export Array ({ path, rule }). "rule" is defined at: <https://github.com/chimurai/http-proxy-middleware>.

  ```javascript
  module.exports = [
    {
      path: "/api",
      rule: { target: "http://www.example.org" },
    },
  ];
  // will be transformed into app.use(path, httpProxyMiddleware(rule))
  ```

- `--cors`: Enable CORS for all requests.
- `--help, -h`: Display usage instructions.

### Inspect

This command can be used to inspect the Webpack config generated by this app extension.

```bash
quasar ssg inspect
```

#### Inspect Options

<details>
  <summary>Vite</summary>

- `--cmd, -c`: Quasar SSG command [dev|generate] (default: dev)
- `-d, --depth`: Number of levels deep (default: 2).
- `-p, --path`: Path of config in dot notation.

  Examples:

  ```bash
  quasar ssg inspect -p build.rollupOptions
  quasar ssg inspect -p plugins
  ```

- `--colors`: Style output with ANSI color codes (default: true).
- `--thread, -t`: Display only one specific ssg config thread
- `-h, --help`: Display usage instructions.
</details>

<details>
  <summary>Webpack</summary>

- `-d, --depth`: Number of levels deep (default: 5).
- `-p, --path`: Path of config in dot notation.

  Examples:

  ```bash
  quasar ssg inspect -p module.rules
  quasar ssg inspect -p plugins
  ```

- `--colors`: Style output with ANSI color codes (default: true).
- `-h, --help`: Display usage instructions.
</details>

## Configuration

Options can be passed with `ssg` key in `/quasar.config.js` file.

```javascript
// quasar.config.js

module.exports = function (/* ctx */) {
  return {
    // ...

    ssg: {
      // pass options here
    },

    // ...
  };
};
```

### Webpack

<details>
  <summary>See all available options</summary>

#### `concurrency`

Type: `Number`

Default: `10`

Page generation is concurrent, `ssg.concurrency` specifies the amount of page generation that runs in one thread.

#### `interval`

Type: `Number`

Default: `0`

Interval in **milliseconds** between two batches of concurrent page generation to avoid flooding a potential API with calls to the API from the web application.

> Notes:
>
> This option is intended to be used in conjunction with the `concurrency` option.
> For example, setting `concurrency` to `10` and `interval` to `5000` will execute the generation of 10 pages in parallel every 5 seconds.

#### `routes`

Type: `String[]` or `Function`

Default: `[]`

A list of routes to generate the corresponding pages.

> Note: As of quasar-app-extension-ssg [v2.0.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v1.2.0...v2.0.0) this option is optionnal due to the crawler feature and the ability to include static routes from the app's router using the `ssg.includeStaticRoutes` option.
>
> If the app has unlinked pages (such as secret pages) and these also need to be generated, the `ssg.routes` property can be used.

Example:

```javascript
ssg: {
  routes: ["/", "/about", "/users", "/users/someone"];
}
```

With a `Function` which returns a `Promise`:

```javascript
// quasar.config.js

const axios = require("axios");

module.exports = function (/* ctx */) {
  return {
    // ...

    ssg: {
      routes() {
        return axios.get("https://my-api/users").then((res) => {
          return res.data.map((user) => {
            return "/users/" + user.id;
          });
        });
      },
    },

    // ...
  };
};
```

With a `Function` which returns a `callback(err, params)`:

```javascript
// quasar.config.js

const axios = require("axios");

module.exports = function (/* ctx */) {
  return {
    // ...

    ssg: {
      routes(callback) {
        axios
          .get("https://my-api/users")
          .then((res) => {
            const routes = res.data.map((user) => {
              return "/users/" + user.id;
            });
            callback(null, routes);
          })
          .catch(callback);
      },
    },

    // ...
  };
};
```

#### `includeStaticRoutes`

:new: _Added in [v4.0.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v3.4.0...v4.0.0)_

Type: `Boolean`

Default: `true`

Include the application's router static routes to generate the corresponding pages.

> Note: In case of warnings issued when initializing routes, this option can be disabled. Then the `crawler` feature and the `ssg.routes` options can be used to provide the static and dynamic routes.

#### `distDir`

:new: _Added in [v4.2.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.1.1...v4.2.0)_

Type: `String`

Default: `'<project-folder>/dist/ssg'`

Folder where the extension should generate the distributables. Relative path to project root directory.

#### `buildDir`

Type: `String`

Default: `'<project-folder>/node_modules/.cache/quasar-app-extension-ssg'` or `'<project-folder>/.ssg-build'` if `cache` is set to false.

The webpack compilation output folder from where the extension can prerender pages.

#### `cache`

Type: `Object` or `false`

Default:

```javascript
{
  ignore: [
    join(conf.ssg.distDir, '/**'), // dist/ssg
    join(conf.ssg.buildDir, '/**'), // node_modules/.cache/quasar-app-extension-ssg
    ...conf.build.distDir ? [join(conf.build.distDir, '/**')] : [],
    'dist/**',
    'public/**',
    'src-ssr/**',
    'src-cordova/**',
    'src-electron/**',
    'src-bex/**',
    'src/ssg.d.ts',
    'node_modules/**',
    '.**/*',
    '.*',
    'README.md'
  ],
  globbyOptions: {
    gitignore: true
  }
}
```

This option caches the compilation output folder and skips recompilation when no tracked file has changed.

- `ignore` is a [Globby](https://github.com/sindresorhus/globby#patterns) patterns to ignore tracked files. If an array is provided, it will be merged with default options. A function can be passed to return an array that will remove the defaults.

  Example with an `Array`:

  ```javascript
  ssg: {
    cache: {
      ignore: ["renovate.json"]; // ignore changes applied on this file
    }
  }
  ```

  With a `Function`:

  ```javascript
  ssg: {
    cache: {
      ignore: (defaultIgnore) =>
        defaultIgnore.push("renovate.json") && defaultIgnore;
    }
  }
  ```

- `globbyOptions` can be used to add [globby options](https://github.com/sindresorhus/globby#options).

#### `fallback`

Type: `String`

Default: `'404.html'`

The name of the SPA/PWA fallback file intended to be served when an index.html file does not exist for a given route.

> Notes:
>
> - Overrides `build.htmlFilename` and `build.ssrPwaHtmlFilename`.
> - This file is created with the webpack plugin [`html-webpack-plugin`](https://github.com/jantimon/html-webpack-plugin) via [this Quasar file](https://github.com/quasarframework/quasar/blob/dev/app-webpack/lib/webpack/inject.html.js). It can be extended with some [plugins](https://github.com/jantimon/html-webpack-plugin#plugins).
> - Multiple services (e.g. Netlify, Vercel) detect a 404.html automatically. For custom web server, an error page must be set up and set to the 404.html file.

#### `crawler`

:new: _Added in [v2.0.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v1.2.0...v2.0.0)_

Type: `Boolean`

Default: `true`

Crawls html links as each page is generated to find dynamic and static routes to add to the page generation queue.

#### `exclude`

:new: _Added in [v2.0.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v1.2.0...v2.0.0)_

Type: `String[] | Regexp[]`

An array of routes or regular expressions matching them to prevent corresponding pages from being generated.

Example with an `Array of String`:

```javascript
ssg: {
  exclude: ["/my-secret-page"];
}
```

With an `Array of Regexp`:

```javascript
ssg: {
  exclude: [
    /^\/admin/, // path starts with /admin
  ];
}
```

#### `shouldPreload(file, type, ext, isLazilyHydrated)`

:new: _Added in [v3.3.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v3.2.1...v3.3.0)_

Type: `Function`

A function to control what files should have <link rel="preload"> resource hints generated.

By default, no assets will be preloaded.

Example to preload assets:

```javascript
ssg: {
  shouldPreload: (file, type, ext) => {
    // type is inferred based on the file extension.
    // https://fetch.spec.whatwg.org/#concept-request-destination
    if (type === "script" || type === "style") {
      return true;
    }
    if (type === "font" && ext === "woff2") {
      // only preload woff2 fonts
      return file;
    }
    if (type === "image") {
      // only preload important images
      return file === "hero.jpg";
    }
  };
}
```

#### `shouldPrefetch(file, type, ext, isLazilyHydrated)`

:new: _Added in [v3.3.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v3.2.1...v3.3.0)_

Type: `Function`

A function to control what files should have <link rel="prefetch"> resource hints generated.

By default no assets will be prefetched. However this is possible to customize what to prefetch in order to better control bandwidth usage. This option expects the same function signature as shouldPreload.

#### `inlineCriticalCss`

Type: `Boolean` or `Object`

Default: `true`

Uses [Beastcss](https://github.com/freddy38510/beastcss) to inline critical CSS and async load the rest for each generated page.

The default beastcss options can be customized by passing them to `inlineCriticalCss`.

Example:

```javascript
ssg: {
  inlineCriticalCss: {
    internal: false,
    merge: false,
  };
}
```

> Notes:
>
> The value is forced to `false` when using the [dev](#dev) command.

#### `inlineCssFromSFC`

:new: _Added in [v3.3.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v3.2.1...v3.3.0)_

Type: `Boolean`

Default: `false`

Inline css from Vue Single-File Component (SFC) `<style>` blocks.

Note: This option works even if build.extractCSS is set to `true` in quasar.config.js file.

> Notes:
>
> The value is forced to `true` when using the [dev](#dev) command.

#### `onRouteRendered(html, route, distDir)`

Type: `Function`

Hook executed after pre-rendering a page just before writing it to the filesystem.

This function must return the html string.

Can use async/await or directly return a Promise.

#### `afterGenerate(files, distDir)`

Type: `Function`

Hook executed after all pages has been generated.

Can use async/await or directly return a Promise.

> Note: The `files` parameter is an `Array` of all generated page paths + filenames (including the fallback file).

</details>

### Vite

<details>
  <summary>See all available options</summary>

#### `concurrency`

Type: `Number`

Default: `10`

Page generation is concurrent, `ssg.concurrency` specifies the amount of page generation that runs in one thread.

#### `interval`

Type: `Number`

Default: `0`

Interval in **milliseconds** between two batches of concurrent page generation to avoid flooding a potential API with calls to the API from the web application.

> Notes:
>
> This option is intended to be used in conjunction with the `concurrency` option.
> For example, setting `concurrency` to `10` and `interval` to `5000` will execute the generation of 10 pages in parallel every 5 seconds.

#### `routes`

Type: `String[]` or `Function`

Default: `[]`

A list of routes to generate the corresponding pages.

> Note: This option is optionnal due to the crawler feature and the ability to include static routes from the app's router using the `ssg.includeStaticRoutes` option.
>
> If the app has unlinked pages (such as secret pages) and these also need to be generated, the `ssg.routes` property can be used.

Example:

```javascript
ssg: {
  routes: ["/", "/about", "/users", "/users/someone"];
}
```

With a `Function` which returns a `Promise`:

```javascript
// quasar.config.js

const axios = require("axios");

module.exports = function (/* ctx */) {
  return {
    // ...

    ssg: {
      routes() {
        return axios.get("https://my-api/users").then((res) => {
          return res.data.map((user) => {
            return "/users/" + user.id;
          });
        });
      },
    },

    // ...
  };
};
```

With a `Function` which returns a `callback(err, params)`:

```javascript
// quasar.config.js

const axios = require("axios");

module.exports = function (/* ctx */) {
  return {
    // ...

    ssg: {
      routes(callback) {
        axios
          .get("https://my-api/users")
          .then((res) => {
            const routes = res.data.map((user) => {
              return "/users/" + user.id;
            });
            callback(null, routes);
          })
          .catch(callback);
      },
    },

    // ...
  };
};
```

#### `includeStaticRoutes`

Type: `Boolean`

Default: `true`

Include the application's router static routes to generate the corresponding pages.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

## [4.7.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.6.0...v4.7.0) (2023-02-14)


### Features

* add support for Vite.js versions 3 and 4 ([d880da5](https://github.com/freddy38510/quasar-app-extension-ssg/commit/d880da5cb8c62a1b7322dfb273c21ff7e806c4f9)), closes [#332](https://github.com/freddy38510/quasar-app-extension-ssg/issues/332)
* **webpack,vite:** improve dependencies handling and ssg extension install/uninstall procedures ([aa95f08](https://github.com/freddy38510/quasar-app-extension-ssg/commit/aa95f0817996aa99096cfcb706fe4d8a8641e0df))


### Bug Fixes

* **deps:** update dependency fs-extra to v11 ([#328](https://github.com/freddy38510/quasar-app-extension-ssg/issues/328)) ([3209c9d](https://github.com/freddy38510/quasar-app-extension-ssg/commit/3209c9dd340be99d3f1cd6a1953432fb432af311))
* **dev:** normalize the requested url before generating the corresponding page ([f523681](https://github.com/freddy38510/quasar-app-extension-ssg/commit/f523681f076f5004b5294ef764262358472e6532)), closes [#321](https://github.com/freddy38510/quasar-app-extension-ssg/issues/321)
* let Vue throw all errors when rendering server-side at build time ([15ac18b](https://github.com/freddy38510/quasar-app-extension-ssg/commit/15ac18be3064e637ddf87c685af1ded8ffcfc39d))
* **package:** add @rollup/plugin-node-resolve to dependencies instead of devDependencies ([9592666](https://github.com/freddy38510/quasar-app-extension-ssg/commit/9592666480af4e57d65abfcb55db4a1ee74ae20e))
* **quasar-config:** ensure the ctx property is set ([88f2f16](https://github.com/freddy38510/quasar-app-extension-ssg/commit/88f2f16af9c3346e6d0db3d839d25317384ffd11))
* **webpack,vite:** do not add ssg-corrections boot file if quasar is greater than v2.11.1 ([13bfa3e](https://github.com/freddy38510/quasar-app-extension-ssg/commit/13bfa3eb639d6649c1916e5dbfc39df50e7f9e3a))
* **webpack,vite:** do not leak currentScript when on SSR with Pinia/Vuex ([43c8aa8](https://github.com/freddy38510/quasar-app-extension-ssg/commit/43c8aa8c5747cec848f36fe4323dc8c485208bcf))
* **webpack:** support css-loader major version 6 since @quasar/app-webpack v3.7.0 update ([4dd9489](https://github.com/freddy38510/quasar-app-extension-ssg/commit/4dd9489780b5e995366c109cc3595086102b7aa3))

## [4.6.0](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.5.8...v4.6.0) (2022-10-14)


### Features

* add ability to provide types for quasar configuration ssg property when installing extension ([0994172](https://github.com/freddy38510/quasar-app-extension-ssg/commit/0994172069654fb6e648991f053c87a1272ee59c))


### Bug Fixes

* **cache:** ignore ssg declaration types file ([38d4e07](https://github.com/freddy38510/quasar-app-extension-ssg/commit/38d4e0796b581f84703993f1c299f0cac0173b07))
* **deps:** update dependency beastcss to v2 ([#311](https://github.com/freddy38510/quasar-app-extension-ssg/issues/311)) ([5989843](https://github.com/freddy38510/quasar-app-extension-ssg/commit/598984325042c1407f3f8e3ccf6b27b572ebf409))
* **deps:** update dependency node-html-parser to v6 ([#312](https://github.com/freddy38510/quasar-app-extension-ssg/issues/312)) ([b725794](https://github.com/freddy38510/quasar-app-extension-ssg/commit/b7257947b8a5eed78ae600982607a0a5921d8e1d))
* **vite,webpack:** handle boot import errors gracefully ([e970291](https://github.com/freddy38510/quasar-app-extension-ssg/commit/e97029163facdaf323b332c40c212d9f4089a18a))
* **vite:** add missing generation of types feature flag ([3e0e605](https://github.com/freddy38510/quasar-app-extension-ssg/commit/3e0e60597b4a0028ebbf27b16af6f7791658711e))
* **vite:** allow dev server to serve files from extension folder ([22b3a1e](https://github.com/freddy38510/quasar-app-extension-ssg/commit/22b3a1ea19f07db0b9182d163aa23d396732253f))
* **vite:** avoid restarting the dev server at launch while developing the extension ([2ce1bf7](https://github.com/freddy38510/quasar-app-extension-ssg/commit/2ce1bf7ea26857021887bd50530b8bc03e111c23))
* **vite:** check entry module before collecting the styles with the dev command ([9345a0f](https://github.com/freddy38510/quasar-app-extension-ssg/commit/9345a0f58e02e166df1154ff8cd31d784e5bdc6e))
* **vite:** handle properly generated html strings with null value ([3bcb383](https://github.com/freddy38510/quasar-app-extension-ssg/commit/3bcb3837cecdab78a082dac41b04ae97b4f016d9))
* **vite:** improve and future-proof the transform checks of ssg vite plugins ([2ac5b2e](https://github.com/freddy38510/quasar-app-extension-ssg/commit/2ac5b2ec44b5a1fc2b6cc2163a7d55a8b90ea0c4))
* **vite:** support linked app-extension-ssg package (via yarn link command) for development purposes ([97720f0](https://github.com/freddy38510/quasar-app-extension-ssg/commit/97720f0cf1bffb1f7bfcec72246062eaced4ce02))

### [4.5.8](https://github.com/freddy38510/quasar-app-extension-ssg/compare/v4.5.7...v4.5.8) (2022-08-18)


### Bug Fixes

* **deps:** update dependency beastcss to v1.2.1 ([#296](https://github.com/freddy38510/quasar-app-extension-ssg/issues/296)) ([e5818b8](https://github.com/freddy38510/quasar-app-extension-ssg/commit/e5818b804f5d9840b512537ce16649356fc2be14))
* **vite:serve:** use built-in https Node.js module instead of obsolete https package ([ba0e8e7](https://github.com/freddy38510/quasar-app-extension-ssg/commit/ba0e8e7d41186c3aab24fc93b1ffd7595458a07b)), closes [#295](https://github.com/freddy38510/quasar-app-extension-ssg/issues/295)
* **vite:** prevent preFetch from throwing error with new vue-router v4.1.0 ([e52470f](https://github.com/freddy38510/quasar-app-extension-ssg/commit/e52470fddaa8972e07f51dcf353cc5b115b1518b))

src/cmd/server.js
/* eslint-disable import/no-dynamic-require */
/* eslint-disable no-console */
/* eslint-disable global-require */

const parseArgs = require('minimist');

const argv = parseArgs(process.argv.slice(4), {
  alias: {
    p: 'port',
    H: 'hostname',
    g: 'gzip',
    s: 'silent',
    colors: 'colors',
    o: 'open',
    c: 'cache',
    cors: 'cors',
    m: 'micro',
    https: 'https',
    C: 'cert',
    K: 'key',
    P: 'proxy',
    h: 'help',
  },
  boolean: ['g', 'https', 'colors', 'h', 'cors'],
  string: ['H', 'C', 'K', 'prefix-path'],
  default: {
    p: process.env.PORT || 4000,
    H: process.env.HOSTNAME || '0.0.0.0',
    g: true,
    c: 24 * 60 * 60,
    m: 1,
    colors: true,
    'prefix-path': '/',
  },
});
if (argv.help) {
  console.log(`
  Description
    Start a HTTP(S) server on a folder.

  Usage
    $ quasar ssg serve [path]
    $ quasar ssg serve . # serve current folder

  Options
    --port, -p              Port to use (default: 4000)
    --hostname, -H          Address to use (default: 0.0.0.0)
    --prefix-path           Create a virtual path prefix (default: /)
    --gzip, -g              Compress content (default: true)
    --silent, -s            Suppress log message
    --colors                Log messages with colors (default: true)
    --open, -o              Open browser window after starting
    --cache, -c <number>    Cache time (max-age) in seconds;
                            Does not apply to /service-worker.js
                            (default: 86400 - 24 hours)
    --micro, -m <seconds>   Use micro-cache (default: 1 second)

    --https                 Enable HTTPS
    --cert, -C [path]       Path to SSL cert file (Optional)
    --key, -K [path]        Path to SSL key file (Optional)
    --proxy <file.js>       Proxy specific requests defined in file;
                            File must export Array ({ path, rule })
                            See example below. "rule" is defined at:
                            https://github.com/chimurai/http-proxy-middleware
    --cors                  Enable CORS for all requests
    --help, -h              Displays this message

  Proxy file example
    module.exports = [
      {
        path: '/api',
        rule: { target: 'http://www.example.org' }
      }
    ]
    --> will be transformed into app.use(path, httpProxyMiddleware(rule))
  `);
  process.exit(0);
}

const fs = require('fs');
const PlatformPath = require('path');
const esmRequire = require('jiti')(__filename);
const express = require('express');
const requireFromApp = require('../helpers/require-from-app');

const { globbySync } = esmRequire('globby');

function getAbsolutePath(pathParam) {
  return PlatformPath.isAbsolute(pathParam)
    ? pathParam
    : PlatformPath.join(process.cwd(), pathParam);
}

const root = getAbsolutePath(argv._[0] || '.');
const resolve = (p) => PlatformPath.resolve(root, p);
const prefixPath = PlatformPath.posix.join('/', argv['prefix-path']);

let green; let grey; let
  red;

if (argv.colors) {
  const chalk = require('chalk');
  green = chalk.green;
  grey = chalk.grey;
  red = chalk.red;
} else {
  red = (text) => text;
  grey = red;
  green = grey;
}

const microCacheSeconds = argv.micro
  ? parseInt(argv.micro, 10)
  : false;

function serve(path, cache) {
  const opts = {
    maxAge: cache ? parseInt(argv.cache, 10) * 1000 : 0,
    setHeaders(res, filePath) {
      if (res.req.method === 'GET' && resolve(filePath).endsWith('.html')) {
        res.set('Cache-Control', 'no-store, no-cache, must-revalidate, proxy-revalidate');
        res.set('Pragma', 'no-cache');
        res.set('Expires', '0');
        res.set('Surrogate-Control', 'no-store');
      }
    },
    extensions: ['html'],
  };

  return express.static(resolve(path), opts);
}

const app = express();

if (argv.cors) {
  const cors = require('cors');
  app.use(cors());
}

if (!argv.silent) {
  app.get('*', (req, _, next) => {
    console.log(
      `GET ${green(req.url)} ${grey(`[${req.ip}]`)} ${new Date()}`,
    );
    next();
  });
}

if (argv.gzip) {
  const compression = require('compression');
  app.use(compression({ threshold: 0 }));
}

const serviceWorkerFile = resolve('service-worker.js');
if (fs.existsSync(serviceWorkerFile)) {
  app.use(PlatformPath.posix.join(prefixPath, 'service-worker.js'), serve('service-worker.js'));
}

if (argv.proxy) {
  argv.proxy = getAbsolutePath(argv.proxy);
  let file = argv.proxy;

  if (!fs.existsSync(file)) {
    console.error(`Proxy definition file not found! ${file}`);
    process.exit(1);
  }
  file = require(file);

  const { createProxyMiddleware } = requireFromApp('http-proxy-middleware');
  file.forEach((entry) => {
    app.use(entry.path, createProxyMiddleware(entry.rule));
  });
}

app.use(prefixPath, serve('.', true));

const fallbackFile = globbySync(resolve('./*.html'), { ignore: [resolve('./index.html')] })[0];

if (fallbackFile) {
  app.use(prefixPath, (req, res, next) => {
    const ext = PlatformPath.posix.extname(req.url) || '.html';

    if (ext !== '.html') {
      return next();
    }

    res.setHeader('Content-Type', 'text/html');

    if (fallbackFile.endsWith('404.html')) {
      res.status(404);

      if (!argv.silent) {
        console.log(red(`  404 on ${req.url}`));
      }
    } else {
      res.status(200);
    }

    return res.sendFile(fallbackFile);
  });
}

if (microCacheSeconds) {
  const microcache = require('route-cache');
  app.use(
    microcache.cacheSeconds(
      microCacheSeconds,
      (req) => req.originalUrl,
    ),
  );
}

app.get('*', (req, res) => {
  res.setHeader('Content-Type', 'text/html');
  res.status(404).send('404 | Page Not Found');
  if (!argv.silent) {
    console.log(red(`  404 on ${req.url}`));
  }
});

function getHostname(host) {
  return host === '0.0.0.0'
    ? 'localhost'
    : host;
}

function getServer() {
  if (!argv.https) {
    return app;
  }

  let fakeCert; let key; let
    cert;

  if (argv.key && argv.cert) {
    key = getAbsolutePath(argv.key);
    cert = getAbsolutePath(argv.cert);

    if (fs.existsSync(key)) {
      key = fs.readFileSync(key);
    } else {
      console.error(`SSL key file not found!${key}`);
      process.exit(1);
    }

    if (fs.existsSync(cert)) {
      cert = fs.readFileSync(cert);
    } else {
      console.error(`SSL cert file not found!${cert}`);
      process.exit(1);
    }
  } else {
    // Use a self-signed certificate if no certificate was configured.
    // Cycle certs every 24 hours
    const certPath = PlatformPath.join(__dirname, '../ssl-server.pem');
    let certExists = fs.existsSync(certPath);

    if (certExists) {
      const certStat = fs.statSync(certPath);
      const certTtl = 1000 * 60 * 60 * 24;
      const now = new Date();

      // cert is more than 30 days old
      if ((now - certStat.ctime) / certTtl > 30) {
        console.log(' SSL Certificate is more than 30 days old. Removing.');
        const { removeSync } = require('fs-extra');
        removeSync(certPath);
        certExists = false;
      }
    }

    if (!certExists) {
      console.log(' Generating self signed SSL Certificate...');
      console.log(' DO NOT use this self-signed certificate in production!');

      const selfsigned = require('selfsigned');
      const pems = selfsigned.generate(
        [{ name: 'commonName', value: 'localhost' }],
        {
          algorithm: 'sha256',
          days: 30,
          keySize: 2048,
          extensions: [{
            name: 'basicConstraints',
            cA: true,
          }, {
            name: 'keyUsage',
            keyCertSign: true,
            digitalSignature: true,
            nonRepudiation: true,
            keyEncipherment: true,
            dataEncipherment: true,
          }, {
            name: 'subjectAltName',
            altNames: [
              {
                // type 2 is DNS
                type: 2,
                value: 'localhost',
              },
              {
                type: 2,
                value: 'localhost.localdomain',
              },
              {
                type: 2,
                value: 'lvh.me',
              },
              {
                type: 2,
                value: '*.lvh.me',
              },
              {
                type: 2,
                value: '[::1]',
              },
              {
                // type 7 is IP
                type: 7,
                ip: '127.0.0.1',
              },
              {
                type: 7,
                ip: 'fe80::1',
              },
            ],
          }],
        },
      );

      try {
        fs.writeFileSync(certPath, pems.private + pems.cert, { encoding: 'utf-8' });
      } catch (err) {
        console.error(` Cannot write certificate file ${certPath}`);
        console.error(' Aborting...');
        process.exit(1);
      }
    }

    fakeCert = fs.readFileSync(certPath);
  }

  return require('https').createServer({
    key: key || fakeCert,
    cert: cert || fakeCert,
  }, app);
}

getServer().listen(argv.port, argv.hostname, () => {
  const url = `http${argv.https ? 's' : ''}://${getHostname(argv.hostname)}:${argv.port}`;
  const fullUrl = url + prefixPath;

  const info = [
    ['Listening at', url],
    prefixPath !== '/' ? ['Served at sub path', fullUrl] : '',
    ['Web server root', root],
    fallbackFile ? ['Fallback', PlatformPath.basename(fallbackFile)] : '',
    argv.https ? ['HTTPS', 'enabled'] : '',
    argv.gzip ? ['Gzip', 'enabled'] : '',
    ['Cache (max-age)', argv.cache || 'disabled'],
    microCacheSeconds ? ['Micro-cache', `${microCacheSeconds}s`] : '',
    argv.cors ? ['CORS', 'enabled'] : '',
    argv.proxy ? ['Proxy definitions', argv.proxy] : '',
  ]
    .filter((msg) => msg)
    .map((msg) => ` ${msg[0].padEnd(20, '.')} ${green(msg[1])}`);

  console.log(`\n${info.join('\n')}\n`);

  if (argv.open) {
    const ci = require('ci-info');

    const isMinimalTerminal = (
      ci.isCI
      || process.env.NODE_ENV === 'test'
      || !process.stdout.isTTY
    );
    if (!isMinimalTerminal) {
      esmRequire('open')(fullUrl);
    }
  }
});
